core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -929,17 +936,17 @@ private[kafka] class Processor(
 
   private object ConnectionId {
     def fromString(s: String): Option[ConnectionId] = s.split("-") match {
-      case Array(local, remote, index) => BrokerEndPoint.parseHostPort(local).flatMap { case (localHost, localPort) =>
+      case Array(local, remote, processorId, index) => BrokerEndPoint.parseHostPort(local).flatMap { case (localHost, localPort) =>
         BrokerEndPoint.parseHostPort(remote).map { case (remoteHost, remotePort) =>
-          ConnectionId(localHost, localPort, remoteHost, remotePort, Integer.parseInt(index))
+          ConnectionId(localHost, localPort, remoteHost, remotePort, Integer.parseInt(processorId), Integer.parseInt(index))
         }
       }
       case _ => None
     }
   }
 
-  private[network] case class ConnectionId(localHost: String, localPort: Int, remoteHost: String, remotePort: Int, index: Int) {
-    override def toString: String = s"$localHost:$localPort-$remoteHost:$remotePort-$index"
+  private[network] case class ConnectionId(localHost: String, localPort: Int, remoteHost: String, remotePort: Int, processorId: Int, index: Int) {
```

Comment:
The connectionId logic is split between `SocketServer` and `Selector.generateConnectionId`. Perhaps we could consolidate them into a top level class ConnectionIdForServer, which can be constructed from either a string or a socket. It also has a `toString` method to get the string representation of the connectionId.

Reply:
I have created following jira: https://issues.apache.org/jira/browse/KAFKA-17813. Working on it, will send the patch. It's a medium refactoring change with tests where BrokerEndpoint is moved as well to java.